magic_firewall_rulesets.go

```diff
@@ -0,0 +1,196 @@
+	Rules       []MagicFirewallRulesetRule `json:"rules"`
+}
+
+// MagicFirewallRulesetRuleActionParameters specifies the action parameters for a Firewall rule
+type MagicFirewallRulesetRuleActionParameters struct {
+	Ruleset string `json:"ruleset,omitempty"`
+}
+
+// MagicFirewallRulesetRule contains information about a single Magic Firewall rule
+type MagicFirewallRulesetRule struct {
+	ID               string                                    `json:"id,omitempty"`
+	Version          string                                    `json:"version,omitempty"`
+	Action           MagicFirewallRulesetRuleAction            `json:"action"`
+	ActionParameters *MagicFirewallRulesetRuleActionParameters `json:"action_parameters,omitempty"`
+	Expression       string                                    `json:"expression"`
+	Description      string                                    `json:"description"`
+	LastUpdated      *time.Time                                `json:"last_updated,omitempty"`
+	Ref              string                                    `json:"ref,omitempty"`
+	Enabled          bool                                      `json:"enabled"`
+}
+
+// CreateMagicFirewallRulesetRequest contains data for a new Firewall ruleset
+type CreateMagicFirewallRulesetRequest struct {
+	Name        string                     `json:"name"`
+	Description string                     `json:"description"`
+	Kind        string                     `json:"kind"`
+	Phase       string                     `json:"phase"`
+	Rules       []MagicFirewallRulesetRule `json:"rules"`
+}
+
+// UpdateMagicFirewallRulesetRequest contains data for a Magic Firewall ruleset update
+type UpdateMagicFirewallRulesetRequest struct {
+	Description string                     `json:"description"`
+	Rules       []MagicFirewallRulesetRule `json:"rules"`
+}
+
+// ListMagicFirewallRulesetResponse contains a list of Magic Firewall rulesets
+type ListMagicFirewallRulesetResponse struct {
+	Response
+	Result []MagicFirewallRuleset `json:"result"`
+}
+
+// GetMagicFirewallRulesetResponse contains a single Magic Firewall Ruleset
+type GetMagicFirewallRulesetResponse struct {
+	Response
+	Result MagicFirewallRuleset `json:"result"`
+}
+
+// CreateMagicFirewallRulesetResponse contains response data when creating a new Magic Firewall ruleset
+type CreateMagicFirewallRulesetResponse struct {
+	Response
+	Result MagicFirewallRuleset `json:"result"`
+}
+
+// UpdateMagicFirewallRulesetResponse contains response data when updating an existing Magic Firewall ruleset
+type UpdateMagicFirewallRulesetResponse struct {
+	Response
+	Result MagicFirewallRuleset `json:"result"`
+}
+
+// ListMagicFirewallRulesets lists all Rulesets for a given account
+//
+// API reference: https://api.cloudflare.com/#rulesets-list-rulesets
+func (api *API) ListMagicFirewallRulesets(ctx context.Context) ([]MagicFirewallRuleset, error) {
+	uri := fmt.Sprintf("/accounts/%s/rulesets", api.AccountID)
```

Comment:
My only hesitation is the lack of validation around usage of `api.AccountID` in a couple of these functions as they are not a part of the signature or validated not empty before being utilised. Within upstream clients (such as the Terraform provider) we can perform a check to speed up the feedback cycle however I do wonder if we are going to make this difficult to debug if someone forgets it as the URL will end up being `/accounts//rulesets` with a routing error from the API gateway. Thoughts?